docs/decisions/0045-updated-vector-store-design.md

```diff
@@ -0,0 +1,749 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
+contact: westey-m
+date: 2024-05-01
+deciders: sergeymenshykh, markwallace, rbarreto, dmytrostruk, westey-m
+consulted: 
+informed: 
+---
+
+# Updated Memory Connector Design
+
+## Context and Problem Statement
+
+Semantic Kernel has a collection of connectors to popular Vector databases e.g. Azure AI Search, Chroma, Milvus, ...
+Each Memory connector implements a memory abstraction defined by Semantic Kernel and allows developers to easily integrate Vector databases into their applications.
+The current abstractions are experimental and the purpose of this ADR is to progress the design of the abstractions so that they can graduate to non experimental status.
+
+### Problems with current design
+
+1. The `IMemoryStore` interface has four responsibilities with different cardinalities.
+2. The `IMemoryStore` interface only supports a fixed schema for data storage, retrieval and search, which limits its usability by customers with existing data sets.
+2. The `IMemoryStore` implementations are opinionated around key encoding / decoding and collection name sanitization, which limits its usability by customers with existing data sets.
+
+Responsibilities:
+
+|Functional Area|Cardinality|Significance to Semantic Kernel|
+|-|-|-|
+|Collection/Index create|An implementation per store type and model|Valuable when building a store and adding data|
+|Collection/Index list/exists/delete|An implementation per store type|Valuable when building a store and adding data|
+|Data Storage and Retrieval|An implementation per store type|Valueble when building a store and adding data|
+|Vector Search|An implementation per store type, model and search type|Valuable for RAG scenarios|
```

Comment:
I would argue this is not just about RAG, but one of the main goals of "memory" as a feature. For instance, search by similarity can be used in these scenarios:
* when receiving user input, search for similar input to understand the User Intent
* when planning for a given goal, search for relevant functions to execute
* when building a prompt with examples (few shots learning), search for relevant examples
* when saving a "memory", search for similar memories to be merged/updated
* when validating user input, search for contradicting facts

Reply:
Makes sense, I've included a few more of the examples here in the doc to clarify.